app/models/course_data/articles_courses.rb

```diff
@@ -64,9 +64,9 @@ def article_revisions
   end
 
   def update_cache
-    revisions = live_manual_revisions
+    revisions = live_manual_revisions.load
 
-    self.character_sum = revisions.where('characters >= 0').sum(:characters)
+    self.character_sum = revisions.select { |r| r.characters >= 0 }.sum(&:characters)
```

Comment:
All of these changes to the cache-updating code go against my incorrect intuitions about what is faster. I've taken a few passes at trying to speed this up before, but usually working from the assumption of 'the database is faster than ruby, so do as much ordering and selecting in the database as you can.

My takeaway is that — even if doing a single calculation in the database is faster — ruby Enumerable operations might be faster than going back to the database multiple times.

Reply:
The critical line is the next one - `revisions.sum(&:references_added)`. Because we have to do this, that means `revisions` _must_ be loaded. So, if we have to load it, just load it earlier and do all of the other calculations in-memory. 

If we _didn't_ need to load all `revisions`, doing orders/filtering in the database would probably be faster.

Reply:
Someone on Twitter pointed out this might be a good point to use window aggregate functions. Maybe next time, I think this Ruby version is fast enough.